sdk/go/pulumi/context.go

```diff
@@ -308,6 +308,182 @@ func (ctx *Context) Invoke(tok string, args interface{}, result interface{}, opt
 	return nil
 }
 
+// Call will invoke a provider call function, identified by its token tok.
+//
+// output is used to determine the output type to return; self is optional for methods.
+func (ctx *Context) Call(tok string, args Input, output Output, self Resource, opts ...InvokeOption) (Output, error) {
+	if tok == "" {
+		return nil, errors.New("call token must not be empty")
+	}
+
+	output = ctx.newOutput(reflect.TypeOf(output))
+
+	options := &invokeOptions{}
+	for _, o := range opts {
+		if o != nil {
+			o.applyInvokeOption(options)
+		}
+	}
+
+	// Note that we're about to make an outstanding RPC request, so that we can rendezvous during shutdown.
+	if err := ctx.beginRPC(); err != nil {
+		return nil, err
+	}
+
+	prepareCallRequest := func() (*pulumirpc.CallRequest, error) {
+		// Determine the provider and version to use.
+		var provider ProviderResource
+		var version string
+		if self != nil {
+			provider = self.getProvider()
+			version = self.getVersion()
+		} else {
+			provider = mergeProviders(tok, options.Parent, options.Provider, nil)[getPackage(tok)]
+			version = options.Version
+		}
+		var providerRef string
+		if provider != nil {
+			pr, err := ctx.resolveProviderReference(provider)
+			if err != nil {
+				return nil, err
+			}
+			providerRef = pr
+		}
+
+		// Serialize all args, first by awaiting them, and then marshaling them to the requisite gRPC values.
+		resolvedArgs, argDeps, _, err := marshalInputs(args)
+		if err != nil {
+			return nil, fmt.Errorf("marshaling args: %w", err)
+		}
+
+		// If we have a value for self, add it to the arguments.
+		if self != nil {
+			var deps []URN
+			resolvedSelf, selfDeps, err := marshalInput(self, nil, true)
+			if err != nil {
+				return nil, fmt.Errorf("marshaling __self__: %w", err)
+			}
+			for _, dep := range selfDeps {
+				depURN, _, _, err := dep.URN().awaitURN(context.TODO())
+				if err != nil {
+					return nil, err
+				}
+				deps = append(deps, depURN)
+			}
+			resolvedArgs["__self__"] = resolvedSelf
+			argDeps["__self__"] = deps
+		}
+
+		// Marshal all properties for the RPC call.
+		keepUnknowns := ctx.DryRun()
+		rpcArgs, err := plugin.MarshalProperties(
+			resolvedArgs,
+			plugin.MarshalOptions{KeepSecrets: true, KeepUnknowns: keepUnknowns, KeepResources: ctx.keepResources})
```

Comment:
I was poking around messy bugs with MarshalOptions recently. I think the code here says this: if ctx.DryRun()=true (preview?), preserve as-is, otherwise try to filter unknowns out (which afaik can change the "shape" of the data by e.g. removing required properties). Is that the desired behavior?

There is also an option to RejectUnknowns=true which will fail with an exception if it sees any unknowns.

What I'm thinking could be an alternative here. If `ctx.DryRun() = false` (we're applying changes), be strict and RejectUnknowns since there should not be any. If `ctx.DryRun() = true` need to choose between `KeepUnknowns=true` and `KeepUnknowns=false` depending on whether the implementing code knows how to handle them. Not knowing much about this,`KeepUnknowns=true` seems like a better default to me since I don't fully trust the filtering process.

Reply:
> ctx.DryRun()=true (preview?)

Yep, `DryRun` means preview.

> What I'm thinking could be an alternative here.

I like the idea of being strict and setting `RejectUnknowns=true` when `DryRun` is false. When `DryRun` is false there should not be any unknown values, so whether we're filtering or strictly rejecting, it shouldn't matter either way, and rejecting would be a good assertion to have.

However, the code here is based on the exact same marshaling options used when registering/reading a resource:

https://github.com/pulumi/pulumi/blob/2c0d6924cbf374931857c17de8d15712d3b6d057/sdk/go/pulumi/context.go#L1155-L1162

My preference is to not introduce `RejectUnknowns` as part of this change, and instead consider switching to using it here (and elsewhere) consistently in a subsequent change.

Reply:
Trying here https://github.com/pulumi/pulumi/pull/7457